File: crawler/src/providers/factory.ts
import BaseProvider from "./base_provider";
import GuwenwenProvider from "./guwenwen_provider";
import MingyanProvider from "./mingyan_provider";
import GithubDatasetProvider from "./github_dataset_provider";
import WikiquoteProvider from "./wikiquote_provider";

const registry: Map<string, () => BaseProvider> = new Map();

registry.set("guwenwen", () => new GuwenwenProvider());
registry.set("mingyan", () => new MingyanProvider());
registry.set("github_dataset", () => new GithubDatasetProvider());
registry.set("wikiquote", () => new WikiquoteProvider());

export function getProvider(code: string): BaseProvider | null {
  const factory = registry.get(code);
  if (!factory) return null;
  return factory();
}

export function listProviders(): string[] {
  return Array.from(registry.keys());
}

export function getAllProviders(): BaseProvider[] {
  return Array.from(registry.values()).map((factory) => factory());
}

export function registerProvider(code: string, factory: () => BaseProvider): void {
  registry.set(code, factory);
}

export default { getProvider, listProviders, getAllProviders, registerProvider };

File: crawler-py/crawler/providers/factory.py
"""Provider 工厂

对应 TypeScript 版 crawler/src/providers/factory.ts
"""
from typing import Callable, Dict, List, Optional

from .base_provider import BaseProvider
from .github_dataset_provider import GithubDatasetProvider
from .guwenwen_provider import GuwenwenProvider
from .mingyan_provider import MingyanProvider
from .wikiquote_provider import WikiquoteProvider

registry: Dict[str, Callable[[], BaseProvider]] = {}

registry["guwenwen"] = GuwenwenProvider
registry["mingyan"] = MingyanProvider
registry["github_dataset"] = GithubDatasetProvider
registry["wikiquote"] = WikiquoteProvider


def get_provider(code: str) -> Optional[BaseProvider]:
    """对应 TS getProvider"""
    factory = registry.get(code)
    if not factory:
        return None
    return factory()


def list_providers() -> List[str]:
    """对应 TS listProviders"""
    return list(registry.keys())


def get_all_providers() -> List[BaseProvider]:
    """对应 TS getAllProviders"""
    return [factory() for factory in registry.values()]


def register_provider(code: str, factory: Callable[[], BaseProvider]) -> None:
    """对应 TS registerProvider"""
    registry[code] = factory
